File: VB/InfiniteAsyncSource.GRPC/IssueFilterConverter.vb
Imports DevExpress.Data.Filtering
Imports DevExpress.Xpf.Data
Imports IssuesData
Imports System
Imports System.Globalization
Imports System.Linq
Imports System.Windows.Data
Imports System.Windows.Markup

Namespace InfiniteAsyncSource.GRPC
	Public Class IssueFilterConverter
		Inherits MarkupExtension
		Implements IValueConverter

		Private Function IValueConverter_Convert(ByVal filter As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.Convert
			Return DirectCast(filter, CriteriaOperator).Match(binary:= Function(propertyName, value, type)
				If propertyName = "Votes" AndAlso type = BinaryOperatorType.GreaterOrEqual Then
					Return New IssueFilter With {.MinVotes = CInt(Math.Truncate(value))}
				End If
				If propertyName = "Priority" AndAlso type = BinaryOperatorType.Equal Then
					Return New IssueFilter With {.Priority = CType(value, Priority)}
				End If
				If propertyName = "Created" Then
					If type = BinaryOperatorType.GreaterOrEqual Then
						Return New IssueFilter With {.CreatedFrom = CDate(value)}
					End If
					If type = BinaryOperatorType.Less Then
						Return New IssueFilter With {.CreatedTo = CDate(value)}
					End If
				End If
				Throw New InvalidOperationException()
			End Function, [and]:= Function(filters)
				Return New IssueFilter With {
					.CreatedFrom = filters.Select(Function(x) x.CreatedFrom).SingleOrDefault(Function(x) x IsNot Nothing),
					.CreatedTo = filters.Select(Function(x) x.CreatedTo).SingleOrDefault(Function(x) x IsNot Nothing),
					.MinVotes = filters.Select(Function(x) x.MinVotes).SingleOrDefault(Function(x) x IsNot Nothing),
					.Priority = filters.Select(Function(x) x.Priority).SingleOrDefault(Function(x) x IsNot Nothing)
				}
			End Function, null:= CType(Nothing, IssueFilter))
		End Function
		Private Function IValueConverter_ConvertBack(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
			Throw New NotImplementedException()
		End Function
		Public Overrides Function ProvideValue(ByVal serviceProvider As IServiceProvider) As Object
			Return Me
		End Function
	End Class
End Namespace

File: CS/InfiniteAsyncSource.GRPC/IssueFilterConverter.cs
using DevExpress.Data.Filtering;
using DevExpress.Xpf.Data;
using IssuesData;
using System;
using System.Globalization;
using System.Linq;
using System.Windows.Data;
using System.Windows.Markup;

namespace InfiniteAsyncSource.GRPC {
    public class IssueFilterConverter : MarkupExtension, IValueConverter {
        object IValueConverter.Convert(object filter, Type targetType, object parameter, CultureInfo culture) {
            return ((CriteriaOperator)filter).Match(
                binary: (propertyName, value, type) => {
                    if(propertyName == "Votes" && type == BinaryOperatorType.GreaterOrEqual)
                        return new IssueFilter { MinVotes = (int)value };

                    if(propertyName == "Priority" && type == BinaryOperatorType.Equal)
                        return new IssueFilter { Priority = (Priority)value };

                    if(propertyName == "Created") {
                        if(type == BinaryOperatorType.GreaterOrEqual)
                            return new IssueFilter { CreatedFrom = (DateTime)value };
                        if(type == BinaryOperatorType.Less)
                            return new IssueFilter { CreatedTo = (DateTime)value };
                    }

                    throw new InvalidOperationException();
                },
                and: filters => {
                    return new IssueFilter {
                        CreatedFrom = filters.Select(x => x.CreatedFrom).SingleOrDefault(x => x != null),
                        CreatedTo = filters.Select(x => x.CreatedTo).SingleOrDefault(x => x != null),
                        MinVotes = filters.Select(x => x.MinVotes).SingleOrDefault(x => x != null),
                        Priority = filters.Select(x => x.Priority).SingleOrDefault(x => x != null)
                    };
                },
                @null: default(IssueFilter)
            );
        }
        object IValueConverter.ConvertBack(object value, Type targetType, object parameter, CultureInfo culture) {
            throw new NotImplementedException();
        }
        public override object ProvideValue(IServiceProvider serviceProvider) => this;
    }
}
